### DISCUSSION TITLE
Alt-Text

### DISCUSSION THREAD

**Discussion by omachado-id** [djplaner/canvas-collections#287] [Ideas]
While on the subject of a question that I just posted, I was reminded to offer these suggestions for a future version of Canvas Collections.

Suggested for the graffiti card/banner link editing area:

A) add an alternate text field so that images added as graffiti cards will be accessible.
B) add the ability to upload graffiti cards from our computers' own file explorer system (to avoid having to link to external images)
C) (this one may require some work...): add the ability to add caption files (e.g,, .SRT, .VTT) to uploaded videos using in the graffiti-card views.

Thanks for considering these suggestions.

**Comment by djplaner** [comment]
G'day,

Thank you again for taking the time to share suggestions.  

## Alt text

Long overdue feature - I've [added an issue](https://github.com/djplaner/canvas-collections/issues/289) to do this. Hopefully sooner, rather than later. 

## Upload card images, rather than external images

The method we've been using for this is to 

1. Upload images to the Canvas course site files area. 
2. Copy and paste the resulting URL

It does rely on a bit of tacit knowledge about Canvas, so I've updated the [documentation for the banner image URL](https://djplaner.github.io/canvas-collections/reference/conceptual-model/objects/banner/#image-url) to include a link to [a description of this process](https://djplaner.github.io/canvas-collections/reference/conceptual-model/objects/url-canvas-files-image/)

## Adding caption files

Given the current model underpinning Collections I think a better solution might be something like the current uploading card image solution.  Following describes that thinking. Mainly as an opportunity to get feedback from you. What do you think? Does this make sense for folk in your context? Would something else work better?

Collections is designed more as a [digital weaver](https://djplaner.github.io/memex/sense/computing/digital-weaver/). A tool to weave together disparate existing strands. e.g the banner component enables weaving together of an existing image or video.  It doesn't provide the hosting of the image or video. That's done somewhere else.

Hence the assumption that the video included in the banner is hosted somewhere else (e.g. Studio, YouTube etc). The assumption being that somewhere else should also provide the captions/transcript.  Canvas Studio makes this a bit easier as I believe the iframe may/should provide the option for  to see both captions and the transcript.  YouTube etc should show captions.  A transcript might need to be linked to from elsewhere (e.g. perhaps a link to a transcript file could be placed at the bottom/top of the descriptive text for a card?)

There are times when this weaving breaks down because the strand isn't available (e.g. image alt text) which is when Collections should provide the functionality.  It's much simpler when it's something as simple as an alt text.  Captions would be a fair bit harder to provide support for.

**Reply by omachado-id** [reply]
Thanks for a thorough and very hopeful response!

Regarding Alt-text: that's a very encouraging development, and I'm glad it was on your radar.

Regarding adding videos vs. links.  I think your solution has plenty of merit.  I suggested adding the option to directly upload a file into the tool, because I wasn't sure that the links to the videos would stay alive once a course is copied from one semester/term to the next.  For instance, videos from Canvas Studio do stay linked even after a course is copied, but I wasn't sure a) they'd stay linked if copied with CC involved in the transfer and b) if links to external videos, (e.g., YouTube) would stay linked after a course copy.  I'll do some testing with this in my course sandboxes to make sure that they do transfer with the other materials.  (For instance, as opposed to one tool we currently use where videos must be reattached to a new course after all other content was transferred; we still get faculty asking for help with that issue.)  I think viable solution is to just put the guidance in your user documentation so that they'll know that it's possible.

Regarding the captions issue: I think I see what you mean (correct me if I'm not getting it).  So images and videos come in as links.  Similarly, a field could be added where the link to a transcript might added (possibly with the ability to view or download that transcript).  So, no reason to code it into a player; just add the ability to access the file somehow?  That might be a workaround.  Like other links, a link to a transcript held in the Canvas course's file area might work.  The caveat would be ensuring that the internal link to the file transfers correctly with other content if the course is copied.  **But I think that, if I'm understanding your approach, it would indeed work without the need for excessive rewriting of the tool's code** Captions are more elegant, but by providing a fallback transcript instead, it may address the thorny issue of providing the accommodation that some viewers might need.  

I think your approach makes sense.  I'll play around with course copying to ensure that (for instance) Canvas Studio and YouTube videos stay linked after a full copy.  I'll respond back here with my results for you.

**Reply by djplaner** [reply]
I fear there is a good chance that Canvas Studio iframes may not migrate correctly with course copy.  If you can confirm that, please let me know.  There may need to be an additional step added in how Collections handle this.

**Reply by omachado-id** [reply]
Hi,

Getting back to you with the results of my course copying test. This will be a bit long.

Method: 
I created a blank destination course in Canvas to receive the course copy from the original course in which I was testing Canvas Collections. Before copying, within the original course, I changed one of the graffiti cards to an internally linked image file (i.e., from the Files area).  I also changed another card to a test video from my Canvas Studio account.  For this, I used a video in one of my STUDIO collections (i.g., not from the top level videos in the main Studio page, but one from a sub-folder, pasting the embed code in the correct area of the Banners section of the card editing page.  I then performed the course copy to the empty target course.

Copying/Importing: 
The course migration (all contents) went as usual, no unexpected hangups or timeouts. Once completed, I went to the Modules view, and was pleasantly surprised by the dialog panel you provide with the Proceed/Cancel/Refresh options.  (Nice touch!).  However, after clicking proceed, I was met with the error message stating that I was editing the same content in another browser or tab.  I was not, as I had editing turned OFF in the other tab that was indeed open showing the original course.  However, even after closing the extra tab, and continuing to attempt to finish the import with the new course, I kept receiving the same "editing in another instance" message.

I tried canceling the import, refreshing the import and nothing worked. However, finally, I reset the target (new course) and re-imported the original's contents again.  This time, I verified that the configuration and output pages were published.  And the clicked proceed again.  This time the import worked and the "successful import" message was displayed. I'm not sure what combination of steps that I took made it work, but it did. (I recommend placing extra bold print in the documentation so that eager beavers like me don't miss the issue of only working in ONE tab and not keeping the source course and the target course open in separate tabs--you'll get peppered with how to fix that).

I did get over this issue, but then found that the Modules view only showed the course card from the first of my four Collections areas.  The others would not show up until I opened the Output page, saved it without changes, and when I returned to the Modules view, everything was there. I don't know where or why it lagged in displaying the cards/layout (possibly because of all the externally linked images needing to refresh?)

Results: 
After overcoming the initial hiccups, however, the original course DID copy completely, including all of the layout settings and cards that I had in the original course. The card file that I had linked from the Files area in the original showed up correctly being linked **_now from the new course's Files area_**.  The video I embedded in one of the card areas, from my Studio account, retained that embed link to Studio in the new course.  Caveat: I used the embed code from the global Studio account, not from the Studio menu item in the course.

Summary: Some of my difficulties may have been of my own making, due to my jumping the gun and not reading the documentation fully regarding the copying process.  I would recommend adding a "tips and tricks" style notation letting users know that the copying is not instantaneous, it may take some time and refreshing of the needed pages, if as I suspect, many of the "threads" that you mentioned before need a bit of time to reestablish.  But, the upshot is that internal links from Files areas migrated correctly. Also, seemingly, did the embed from Canvas Studio. (I think it WILL carry over properly, i.e., not break or improperly tie back to their original course. As you point out though, you might want to check that bit of the code to ensure proper migration.  So no need to for those to be re-linked. And, again, the import status dialog panel is a great touch!

Sorry for the long recap, but maybe this will be useful to you.

**Reply by djplaner** [reply]
Thank you for the detailed description.  Apologies for the delay in replying. It's been a busy week.

From what I've read there are a few moving parts in what you describe.  I'm hoping that unpicking those parts and focusing on them separately will help us get a good shared understanding of what happened which we can use to inform future improvements.

Some details follow, but I've created [an issue](https://github.com/djplaner/canvas-collections/issues/291) for me to explore/fix the edit issue you faced.  

## Editing in another browser

If this were a "pure" editing related, it will be [because of this situation](https://djplaner.github.io/canvas-collections/reference/how-tos/failed-to-turn-editing-on/).  Even looking at that page I see that the screen capture isn't showing a decent link to help document.  Will explore that.

As I know think about it, I wonder if this has come about due to a combination of unique factors.  In part related to how we expanded the "lock" hold time.  i.e. I think the issue may be that when you copied the course, you had the edit lock in the source course.  The Canvas copy process has then copied the "edit lock" page to the destination course.  Making Collections think you were editing the destination course in another tab. 

This is a potential hole in Collections code.  Hence the [issue](https://github.com/djplaner/canvas-collections/issues/291) I've created to fix it.

## Output pages and slow update to modules view

I just wanted to clarify this point.

In Collections, [output page](https://djplaner.github.io/canvas-collections/reference/conceptual-model/collections/existing-collections/#output-page) has a very specific meaning. Output pages are associated with Claytons Collection which generates a static copy of HTML placed into a Canvas page (or multiple).  Post course copy, Claytons Collections will need to be manually updated.  I've updated [the documentation](https://djplaner.github.io/canvas-collections/reference/lifecycle/course-copy/overview/#considerations-when-importing) to make this more explicit.

What you see on the Modules page using live Collections after the course copy process should require no more work.  It should be 100% updated. If not, there's something else going on here.  

## Course based studio videos

It was great to hear that your experience with Studio videos worked.  I've had some limited experience with Studio videos. What I'm wondering is if the "location" of the video in Studio will make a difference with course copy.  i.e. if a Studio video was in the course, not your personal library, what impact that might have.

I'll add that to my list of things to check.

Once again, thank you for taking the time to test out Collections and provide feedback.  Much appreciated.

**Reply by omachado-id** [reply]
I'm glad that the test I ran was useful.  And you could be right with regard to the version of the code I'm on.  According to Tampermonkey, I'm currently on v 1.2.1, and clicking the update option in Tampermonkey tells me that I'm up-to-date with the script.  So I may need to fiddle with that end of things so that I do indeed have the latest version, as my 1.2.1 doesn't show the alt-text enhancement that you added.

Your point on course based studio videos is yet another area to test, as I used a video from my global Studio account.  When time permits, I'll set up another test with this in mind and see if it works too.

**Reply by djplaner** [reply]
It appears that the auto update of the userscript version isn't working.  Initially, due to an oversight of mine and now with that fixed I'm still having problems getting the auto update to work.

A 1.2.2 version has been released (part of the oversight to be fixed), the interim solution would be to [reinstall the userscript (click here)](https://github.com/djplaner/canvas-collections/raw/main/dist/canvas-collections.user.js)

**Reply by omachado-id** [reply]
Thanks!  The update seems to have taken effect, David.  Yesterday (late 9/14) I was showing off a course to an assistant dean, in which I used Collections to organize a resource site for the faculty's use--he was quite impressed with Collections, btw-- and I noticed that the version number had updated to 1.2.2.  So you resolved it; no need to reinstall the script!

Now I have to go back in and add the alt-texts; but I'll happily do that.

**Reply by omachado-id** [reply]
Hi, just following up with results of my test of course-based Studio videos being able to transfer to a new course where Canvas Collections was used in the original (i.e., "source" course).  **Unfortunately, I have not found a way to make this work, but there are workarounds**:  

The problem, IMO, is with Studio, not with Canvas Collections: Whereas one is able to obtain a shareable link to a video that can be embedded in course graffiti cards when it's located in a global Studio account, _**no shareable link can be copied for a course-based Studio video**_.  The Instructure folk have not provided that option for course-based videos (at least not yet).  Therefore course-based videos can't be inserted in graffiti cards via shareable links from the course Studio library.

WORKAROUNDS:
The workaround is simply to download the course-based Studio video and place it in the original course's Files area, then copy THAT link into the graffiti card before exportation to a new course.  That will work. The downside of this approach is that it bloats the course contents file before it's imported into the new course.  Alternatively, a course-based Studio video could be uploaded into one's _global_ Studio libraries to obtain a shareable link from there.  

Both of these workarounds involve downloading/uploading the videos into some location from which a shareable link can be obtained that will work after course copying (i.e., either from global Studio libraries or from videos housed in the Files area of the course.

**Reply by djplaner** [reply]
Thank you for the detailed follow up.

This matches what I assumed might be the case.  Canvas hasn't quite resolved all the use cases around Studio videos and courses. The difference in how standard course content (e.g. the files you mentioned) and Studio videos are treated creating differences in how they act with tasks like course copies.

**Comment by djplaner** [comment]
Support for alt text for banner images has now been added - including [addition to documentation](https://djplaner.github.io/canvas-collections/reference/conceptual-model/objects/banner/#image-alternative-text).

Will require an update to the Collections code you're using.

**Reply by omachado-id** [reply]
Thanks so much for your prompt work; that's excellent!  I'll update it on my end.  I'll post results of my course copying test (the good, the bad and the good!) in a moment.
